Flutter fixup

mkg20001 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Packaging some more things for flutter to check and smooth out the raw edges of mkFlutterApp

If yov've got any comments regarding #147107 this is the place to comment

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mkg20001 committed:
firmware-updater: fix license

mkg20001 committed:
flutter.mkFlutterApp: disable fetch buildPhase

Otherwise it tries to run some other commands, this prevents that

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

mkg20001 committed:
duckscript: init at 0.8.10

mkg20001 committed:
flutter: 2.10.0 -> 2.10.1

mkg20001 committed:
firmware-updater: update vendor hash

mkg20001 committed:
fluffychat: fix olm

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

mkg20001 committed:
flutter.mkFlutterApp: allow extra fetch commands

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/fluffychat/default.nix:
```diff
@@ -21,6 +21,20 @@ flutter.mkFlutterApp rec {
 
+  flutterExtraFetchCommands = ''
+    M=$(echo $TMP/.pub-cache/hosted/pub.dartlang.org/matrix-*)
+    sed "s|/usr/lib/x86_64-linux-gnu/libolm.so.3|${olm}/lib/libolm.so.3|g" -i $M/scripts/prepare.sh
+    sed "s|if which flutter >/dev/null; then|exit; if which flutter >/dev/null; then|g" -i $M/scripts/prepare.sh
```
```suggestion
    sed -i $M/scripts/prepare.sh \
      -e "s|/usr/lib/x86_64-linux-gnu/libolm.so.3|${olm}/lib/libolm.so.3|g"  \
      -e "s|if which flutter >/dev/null; then|exit; if which flutter >/dev/null; then|g"
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/fluffychat/default.nix:
```diff
@@ -21,6 +21,20 @@ flutter.mkFlutterApp rec {
+    sed "s|/usr/lib/x86_64-linux-gnu/libolm.so.3|${olm}/lib/libolm.so.3|g" -i $M/scripts/prepare.sh
+    sed "s|if which flutter >/dev/null; then|exit; if which flutter >/dev/null; then|g" -i $M/scripts/prepare.sh
+
+    # debug
+    # cat $M/scripts/prepare.sh
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/fluffychat/default.nix:
```diff
@@ -21,6 +21,20 @@ flutter.mkFlutterApp rec {
+    # cat $M/scripts/prepare.sh
+
+    pushd $M
+    chmod +x ./scripts/*.sh
+    ./scripts/prepare.sh
```
If this is only done to execute one script we can also do

```suggestion
    bash scripts/prepare.sh
```

SuperSandro2000 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -63,6 +63,9 @@ let
     ];
 
+    # avoid pub phase
+    buildPhase = "true";
```
```suggestion
    dontBuild = true;
```

SuperSandro2000 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -76,6 +79,7 @@ let
       flutter config --enable-linux-desktop
       flutter packages get
       flutter build linux || true # so it downloads tools
+      ${if args ? flutterExtraFetchCommands then args.flutterExtraFetchCommands else ""}
```
```suggestion
      ${lib.optionalString args ? flutterExtraFetchCommands args.flutterExtraFetchCommands}
```

SuperSandro2000 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -223,20 +227,25 @@ let
 
-  buildPhase = ''
-    runHook preBuild
+  preConfigurePhases = [ "flutterConfigurePhase" ];
 
+  # avoid cmake phase
+  configurePhase = "true";
+
+  flutterBuildPhase = ''
     # for some reason fluffychat build breaks without this - seems file gets overriden by some tool
     mv pubspec-backup pubspec.yaml
     mkdir -p build/flutter_assets/fonts
 
     flutter packages get --offline -v
     flutter build linux --release -v
+  '';
 
+  buildPhase = ''
+    runHook preBuild
+    runHook flutterBuildPhase
     runHook postBuild
   '';
```
What is this? Why not just use buildPhase with preHooks? Looks to me a lot easier overhaul.

mkg20001 replied:
How do you mean with prehooks? The idea is the user is able to overwrite buildPhase but can optionally call the original one somewhere in the middle of it. What would be the best way to accomplish this?

SuperSandro2000 replied:
If you want to run something before you would put it into preBuild and for after into postBuild. Basically this would already achieve this:

```suggestion
  preConfigurePhases = [ "flutterConfigurePhase" ];

  # avoid cmake phase
  configurePhase = "true";

  buildPhase = ''
    runHook preBuild

    # for some reason fluffychat build breaks without this - seems file gets overriden by some tool
    mv pubspec-backup pubspec.yaml
    mkdir -p build/flutter_assets/fonts

    flutter packages get --offline -v
    flutter build linux --release -v

    runHook postBuild
  '';
```

mkg20001 replied:
Hrm I'll leave the phases as-is for now. It was initially added to package appflowy, but that is already packaged as bin derivation.

SuperSandro2000 commented on pkgs/build-support/flutter/default.nix:
```diff
@@ -223,20 +227,25 @@ let
+  preConfigurePhases = [ "flutterConfigurePhase" ];
 
+  # avoid cmake phase
+  configurePhase = "true";
```
```suggestion
  dontConfigure = true;
```

mkg20001 replied:
The user might want to overwrite configurePhase and then it would be skipped. Also using preConfigurePhases not sure if this will work.

SuperSandro2000 replied:
What happens if we just leave configurePhase and don't do anything to it? Maybe we just want to disable cmake and ninja?

mkg20001 committed:
flutter.mkFlutterApp: fix installing .desktop

Previously this was pulling from $built, which got moved to $out/app,
so the glob didn't do anything. Now uses find on $out/app

mkg20001 committed:
fluffychat: add .desktop

mkg20001 committed:
fluffychat: fix olm FOD
